### PR TITLE
feat(cli): add auto-pagination and streaming output to kubectl ka get

### LINKED ISSUE #2002: CLI: add auto-pagination support to kubectl ka get
## Problem

When querying large datasets, `kubectl ka get` returns at most one page of results (100 by default, up to 1000) and prints a suggested command to fetch the next page. Users must manually copy-paste that command for each batch, which is impractical when there are thousands of resources.

Other CLIs solve this well. For example, GitHub CLI provides `gh api --paginate` for automatic pagination.

## Proposed Solution

Decouple the `--limit` flag from the server-side page size. Currently `--limit` controls both the total number of records returned and the per-request page size, capped at 1000. Instead:

- `--limit N` (where N can exceed 1000) means "return N records total." The CLI auto-paginates internally, making multiple requests as needed.
- `--limit 0` means "fetch all records."
- The per-request page size is derived automatically: `min(limit, 1000)`, or 1000 when limit is 0.

The CLI uses the server's existing `continue` token for cursor-based pagination between requests. No server-side changes are required.

### Examples

```bash
# Fetch the first 500 resources (5 requests of 100)
kubectl ka get pipelineruns.v1.tekton.dev --limit 500

# Fetch all resources
kubectl ka get pipelineruns.v1.tekton.dev --limit 0
```

When `--limit 0` is used and all data is fetched, no pagination message is printed. When `--limit N` stops at the requested count with more data available, the existing pagination message is still shown.

### PR DESCRIPTION
**Pull request by lcarva**
The --limit flag now controls total desired records rather than per-request page size. The CLI auto-paginates using the server's continue token, fetching multiple pages as needed. --limit 0 means "fetch all records." A new --page-size flag independently controls the per-request page size.

Implement watermark-based windowed merge to reduce memory usage when fetching large result sets. Instead of buffering all KubeArchive pages before printing, the CLI processes one page at a time, emitting settled results as watermarks advance. Memory is bounded to Kubernetes resources plus one KubeArchive page rather than the full archive.

All output formats (table, JSON, YAML) stream incrementally through a resourceStreamer interface, flushing results per batch for progressive display.

## Which Issue(s) this Pull Request Resolves

Resolves #2002

## Release Note

```release-note
action required: The `kubectl ka get` command now automatically paginates through the KubeArchive API, allowing retrieval of more than 1000 resources. Results are streamed to the terminal as they arrive. Use `--limit 0` to fetch all matching resources, and `--page-size` to control the number of resources per API request. The meaning of `--limit` has changed: it now caps the total number of results returned and accepts `0` (all) and values greater than 1000, whereas those values were previously rejected. Scripts that relied on the CLI rejecting `--limit 0` or `--limit` above 1000 will now see them succeed.
```

## Notes for Reviewers

- **Auto-pagination**: The CLI now transparently follows `continue` tokens from the KubeArchive API, fetching multiple pages until the requested limit is reached or all results are returned.
- **Streaming output**: Results are printed incrementally as pages arrive rather than buffering everything in memory. This applies to table, JSON, and YAML output formats.
- **`--limit 0`**: A limit of 0 now means "fetch all" instead of being invalid. The previous max of 1000 is removed since pagination handles larger result sets.
- **`--page-size` flag**: New flag to control how many resources are requested per API call (default: `min(limit, 1000)`, max: 1000).
- **Simplified merge logic**: `insertSortedResourcesOptimized` is replaced by `mergeSortedResources` which no longer handles trimming inline. Limit enforcement is now done at the streaming layer.
- **Watermark-based coordination**: When both in-cluster and archived sources are used, a watermark mechanism ensures resources are emitted in correct sorted order across interleaved pages.

### CODE REVIEW

File: pkg/cmd/get.go [hunk 1]
```diff
@@ -435,207 +736,132 @@ func (o *GetOptions) Run() error {
...
+				continue
+			}
+
+			watermark := extractWatermark(pageResources)
+
+			above, below := splitByWatermark(remaining, watermark)
+			remaining = below
+
+			batch := mergeSortedResources(nil, pageResources, false)
+			batch = mergeSortedResources(batch, above, true)
```
[the remark this answers is not in the document]
Comment (lcarva) [reply pkg/cmd/get.go:758]: This seems like an edge case. Not sure if it's worth the added complexity to handle it.

File: pkg/cmd/get.go [hunk 2]
```diff
@@ -324,6 +327,226 @@ func cmpResource(existing *ResourceWithAvailability, target *unstructured.Unstru
...
+		}
+
+		if js.emittedCount > 0 {
+			fmt.Fprint(js.writer, ",\n")
+		}
+
+		itemBytes, err := json.MarshalIndent(rwa.Resource.Object, "        ", "    ")
+		if err != nil {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (lcarva) [reply pkg/cmd/get.go]: It is unlikely that we get a malformed yaml/json but this seems simple enough to just fix it.

[the remark this answers is not in the document]
Comment (lcarva) [reply pkg/cmd/get.go]: Fixed it.

**Comment by lcarva** [comment]
I think this needs an `/ok-to-test` and some GH workflow approval. I had to amend the commit to include the sign-off text which was initially missing.

File: pkg/cmd/get.go [hunk 3]
```diff
@@ -435,207 +736,132 @@ func (o *GetOptions) Run() error {
...
-		if o.Name != "" {
-			if o.InCluster && o.Archived {
-				return fmt.Errorf("resource not found in Kubernetes or KubeArchive")
-			} else if o.InCluster {
-				return fmt.Errorf("resource not found in Kubernetes cluster")
-			} else if o.Archived {
-				return fmt.Errorf("resource not found in KubeArchive")
+			lastContinueToken := getContinueToken(bodyBytes)
+
+			if len(pageResources) == 0 {
```
Comment (maruiz93) [inline pkg/cmd/get.go:744]: **[Medium]** The pagination loop has no maximum iteration guard. When `--limit 0` and the server keeps returning non-empty continue tokens with empty pages (lines 744-749), the CLI loops indefinitely — that code path bypasses the streamer emit/limit logic entirely. Consider adding a `maxPaginationPages` safety cap (e.g., 10000) that breaks with a warning when exceeded.

File: pkg/cmd/get.go [hunk 4]
```diff
@@ -324,6 +327,226 @@ func cmpResource(existing *ResourceWithAvailability, target *unstructured.Unstru
 	return strings.Compare(target.GetName(), existing.Resource.GetName())
```
Comment (maruiz93) [inline pkg/cmd/get.go:327]: **[Medium]** `get.go` has grown to 867 lines — the largest file in `pkg/cmd/`. The `resourceStreamer` interface and its three implementations (table, JSON, YAML) are self-contained and only depend on `limit` — they could live in a dedicated file (e.g., `streamers.go`) or package. Same for the watermark/merge functions (`extractWatermark`, `splitByWatermark`, `mergeSortedResources`, `cmpResource`). Not blocking, but worth considering as a follow-up to keep the file manageable.

File: pkg/cmd/get.go [hunk 5]
```diff
@@ -435,207 +736,132 @@ func (o *GetOptions) Run() error {
...
+
+			above, below := splitByWatermark(remaining, watermark)
+			remaining = below
+
+			batch := mergeSortedResources(nil, pageResources, false)
+			batch = mergeSortedResources(batch, above, true)
+
+			emitted := streamer.emitBatch(batch)
+			if emitted < len(batch) {
+				for _, r := range batch[emitted:] {
```
Comment (maruiz93) [inline pkg/cmd/get.go:762]: **[Medium]** The `moreInCluster` flag isn't set when the limit is hit before all K8s resources in `remaining` are emitted. After the loop breaks at line 770, `remaining` may still have items, but `moreInCluster` stays false — the pagination message would suggest `--archived` only, missing the K8s resources.

The final block (lines 793-798) does handle `remaining`, but only runs after the archived loop ends. If the limit was already reached inside the loop, those remaining K8s resources get emitted but the `moreInCluster` flag was already determined by the batch trim logic, which only saw the current batch — not what's left in `remaining`.

Consider a test case like:
```go
// K8s has resources at 10:05, 10:03, 10:00
// KubeArchive page1 has resources at 10:04, 10:02 (with continue token)
// --limit 3
// Expected: 3 resources emitted, pagination message includes --in-cluster
// because K8s resource at 10:00 was not emitted
```

File: pkg/cmd/get.go [hunk 6]
```diff
@@ -148,6 +155,16 @@ func (o *GetOptions) buildKubeArchiveQueryParams() url.Values {
 	return params
```
Comment (maruiz93) [inline pkg/cmd/get.go:155]: **[Low]** `pageSize()` re-derives the effective page size on every call (3 branches of conditional logic), but the result is always the same after `Complete()` runs. The rest of the codebase resolves flag values once in `Complete()` — e.g., `kubearchiveQueryParams` at line 275. Consider resolving the page size in `Complete()` and storing it in a private field to follow the same pattern.

File: pkg/cmd/get.go [hunk 7]
```diff
@@ -255,19 +277,17 @@ func (o *GetOptions) Complete(flags *pflag.FlagSet, args []string) error {
 	return nil
 }
 
-// insertSortedResourcesOptimized inserts resources into a sorted slice with deduplication and limit handling
-func (o *GetOptions) insertSortedResourcesOptimized(
+// mergeSortedResources merges new resources into an existing sorted slice with
```
Comment (maruiz93) [inline pkg/cmd/get.go:280]: **[Low]** The function comment says "merges new resources into an existing sorted slice" but doesn't specify the expected sort order. `cmpResource` expects newest-first — if a caller passes oldest-first data, the binary search silently finds wrong insertion points. Consider documenting the invariant: "Both `init` and `resources` must be sorted newest-first by creation timestamp."

File: pkg/cmd/get.go [hunk 8]
```diff
@@ -285,27 +305,10 @@ func (o *GetOptions) insertSortedResourcesOptimized(
...
-			}
-			copy(result[idx+1:], result[idx:])
-			result[idx] = resourceWithAvailability
-
-		} else {
-			// Not inserted, so resources are trimmed
-			k8sTrimmed = k8sTrimmed || fromK8s
-			k9eTrimmed = k9eTrimmed || !fromK8s
-		}
+		result = slices.Insert(result, idx, resourceWithAvailability)
```
Comment (maruiz93) [inline pkg/cmd/get.go:308]: **[Medium]** `mergeSortedResources` uses `slices.BinarySearchFunc` + `slices.Insert` inside a loop. The binary search finds the insertion point in O(log n), but `slices.Insert` shifts all subsequent elements — O(n) per call — so the overall loop is O(m×n). The binary search advantage is negated because Go slices don't support O(1) positional insertion; the shift always dominates.

Since both inputs are always sorted newest-first (API results and K8s resources sorted at line 695), a two-pointer merge achieves O(m+n) — including the single-element case, which is O(N) either way (binary search + shift vs. linear walk + append). The per-page batch sizes keep this manageable today, but with auto-pagination removing the 1000-item cap, larger result sets could hit this path more heavily.

Reference for efficiently merging sorted slices in Go: https://codesignal.com/learn/courses/interview-practice-with-classic-coding-questions-in-go/lessons/advanced-slice-manipulation-and-merging-in-go

**Review by maruiz93 — CHANGES_REQUESTED** [review]
Added some comments that I think could improve the PR. Feel free to open follow-up issues if you think it's too much for this pr (specially for the merge of sorted slices change)

**Comment by maruiz93** [comment]
Unit tests need some adjustments to the optimized_sorting - newest_first test. See https://github.com/kubearchive/kubearchive/actions/runs/30466455212/job/90826035289?pr=2010#step:7:3045

Test source: https://github.com/lcarva/kubearchive/blob/fe84b0b5253c6f48247b6846e9c5fef55eb87222/pkg/cmd/get_test.go#L852

**Comment by lcarva** [comment]
@maruiz93, would you like me to address the fullsend comments? Medium, Low, or both categories?

File: pkg/cmd/get.go [hunk 9]
```diff
@@ -107,7 +113,8 @@ func NewGetCmd() *cobra.Command {
 	cmd.Flags().StringVarP(&o.LabelSelector, "selector", "l", o.LabelSelector, "Selector (label query) to filter on, supports '=', '==', '!=', 'in', 'notin'.(e.g. -l key1=value1,key2=value2,key3 in (value3)). Matching objects must satisfy all of the specified label constraints.")
 	cmd.Flags().BoolVar(&o.InCluster, "in-cluster", true, "Include resources from the Kubernetes cluster.")
 	cmd.Flags().BoolVar(&o.Archived, "archived", true, "Include resources from KubeArchive.")
-	cmd.Flags().IntVar(&o.Limit, "limit", o.Limit, "Maximum number of resources to return (default 100, max 1000).")
+	cmd.Flags().IntVar(&o.Limit, "limit", o.Limit, "Maximum number of resources to return (default 100, 0 for all).")
```
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:116]: I think this is a fair note, `--limit` is mentioned in the release notes but it's not mentioned that its expected usage has changed

File: pkg/cmd/get.go [hunk 10]
```diff
@@ -324,6 +327,226 @@ func cmpResource(existing *ResourceWithAvailability, target *unstructured.Unstru
...
+			break
+		}
+
+		if !js.headerDone {
+			fmt.Fprint(js.writer, "{\n    \"apiVersion\": \"v1\",\n    \"items\": [\n")
+			js.headerDone = true
+		}
+
+		itemBytes, err := json.MarshalIndent(rwa.Resource.Object, "        ", "    ")
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:450]: Low effort and provides a better UX, worth addressing

**Comment by lcarva** [comment]
> Unit tests need some adjustments to the optimized_sorting - newest_first test. See https://github.com/kubearchive/kubearchive/actions/runs/30466455212/job/90826035289?pr=2010#step:7:3045
> 
> Test source: https://github.com/lcarva/kubearchive/blob/fe84b0b5253c6f48247b6846e9c5fef55eb87222/pkg/cmd/get_test.go#L852

Ah this is because of the streaming changes. IIUC, the test fetches the items in batch sizes of 1. So two items means two batches. The first entry on the table has a longer name than the second one. So when the second batch is flushed out, the shorter name gives us a narrower column. The test fails the output comparison:

```diff
 NAME              IN-CLUSTER   ARCHIVED   AGE
 new-archive-pod   no           yes        5m
-old-k8s-pod       yes          no         15m
+old-k8s-pod   yes          no         15m
```

This is a limitation of the current approach. The two options I can think of are:

1. Leave it as is and accept that the output will look like that in some cases.
2. Implement an approach where a column width can never shrink by remembering the previous widest width. This is not supported in tabwriter though. We would have to implement the column padding ourselves, which is not too bad.

In either case, preventing the opposite, column width widens between batches, is impossible to solve with a streaming approach.

File: pkg/cmd/get.go [hunk 11]
```diff
@@ -351,81 +574,159 @@ func (o *GetOptions) parseResourcesFromBytes(bodyBytes []byte) ([]*unstructured.
...
 		}
 
-		bodyBytes, apiErr := o.GetFromAPI(KubeArchive, kubearchiveAPIPath)
-		if apiErr != nil {
-			// If KubeArchive fails with authentication error, don't fall back to just Kubernetes
-			if apiErr.StatusCode == http.StatusUnauthorized ||
-				strings.Contains(apiErr.Message, "empty authorization bearer token given") ||
-				strings.Contains(apiErr.Message, "authentication failed") {
-				return fmt.Errorf("KubeArchive authentication required: %s", apiErr.Message)
+		for {
```
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:710]: worth addressing this

File: pkg/cmd/get.go [hunk 12]
```diff
@@ -324,6 +327,226 @@ func cmpResource(existing *ResourceWithAvailability, target *unstructured.Unstru
...
+		archived := "no"
+		if rwa.Archived {
+			archived = "yes"
+		}
+		age := "<unknown>"
+		if !obj.GetCreationTimestamp().Time.IsZero() {
+			age = duration.HumanDuration(time.Since(obj.GetCreationTimestamp().Time))
+		}
+
+		fmt.Fprintf(ts.writer, "%s\t%-10s\t%-8s\t%s\n", obj.GetName(), inCluster, archived, age)
```
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:406]: feel free to ignore this

File: pkg/cmd/get.go [hunk 13]
```diff
@@ -255,19 +277,17 @@ func (o *GetOptions) Complete(flags *pflag.FlagSet, args []string) error {
 	return nil
 }
 
-// insertSortedResourcesOptimized inserts resources into a sorted slice with deduplication and limit handling
-func (o *GetOptions) insertSortedResourcesOptimized(
+// mergeSortedResources merges new resources into an existing sorted slice with
+// deduplication by UID. No limit is applied.
+func mergeSortedResources(
```
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:282]: feel free to ignore this

[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:282]: feel free to ignore this

**Comment by maruiz93** [comment]
> > Unit tests need some adjustments to the optimized_sorting - newest_first test. See https://github.com/kubearchive/kubearchive/actions/runs/30466455212/job/90826035289?pr=2010#step:7:3045
> > Test source: https://github.com/lcarva/kubearchive/blob/fe84b0b5253c6f48247b6846e9c5fef55eb87222/pkg/cmd/get_test.go#L852
> 
> Ah this is because of the streaming changes. IIUC, the test fetches the items in batch sizes of 1. So two items means two batches. The first entry on the table has a longer name than the second one. So when the second batch is flushed out, the shorter name gives us a narrower column. The test fails the output comparison:
> 
> ```diff
>  NAME              IN-CLUSTER   ARCHIVED   AGE
>  new-archive-pod   no           yes        5m
> -old-k8s-pod       yes          no         15m
> +old-k8s-pod   yes          no         15m
> ```
> 
> This is a limitation of the current approach. The two options I can think of are:
> 
> 1. Leave it as is and accept that the output will look like that in some cases.
> 2. Implement an approach where a column width can never shrink by remembering the previous widest width. This is not supported in tabwriter though. We would have to implement the column padding ourselves, which is not too bad.
> 
> In either case, preventing the opposite, column width widens between batches, is impossible to solve with a streaming approach.

right, we could try option 2, but if you want to can create a follow-up issue for this and modify the test so we test the order but not the format of the spaces

File: pkg/cmd/get.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:758]: Agreed, this is an edge case. I'd suggest adding a brief comment near the watermark logic (around `extractWatermark`/`splitByWatermark`) noting that the watermark is timestamp-only, so if a page boundary splits resources sharing the same `creationTimestamp`, ordering or dedup could be slightly off. That way the limitation is documented and easy to find if someone hits it down the road.

**Comment by maruiz93** [comment]
Answered in the bots threads so you can see which ones should be addressed and which ones not

**Comment by lcarva** [comment]
@maruiz93, I believe I addressed your comments and fullsend's.

File: pkg/cmd/get.go (hunk 5 above)
Comment (maruiz93) [reply pkg/cmd/get.go:814]: Addressed in a4278d8 — both the in-loop path (lines 761–766) and the post-loop path (lines 792–795) now set `moreInCluster` when the limit is hit before all K8s resources in `remaining` are emitted.

File: pkg/cmd/get.go (hunk 3 above)
Comment (maruiz93) [reply pkg/cmd/get.go:793]: Addressed in e839f35 — added `maxPaginationPages = 10000` safety cap. The loop now checks `pagesFetched >= maxPaginationPages`, prints a warning, and breaks.

File: pkg/cmd/get.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go]: Addressed in e839f35 — both JSON and YAML streamer `emitBatch` methods now return `fmt.Errorf(...)` instead of silently continuing on marshal failure.

File: pkg/cmd/get.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (maruiz93) [reply pkg/cmd/get.go:752]: Addressed in e839f35 — same fix as the pagination guard: `maxPaginationPages = 10000` cap with warning and break.

File: pkg/cmd/get.go (hunk 8 above)
Comment (maruiz93) [reply pkg/cmd/get.go:333]: Tracked in #2093 for a follow-up.

**Review by maruiz93 — APPROVED** [review]

**Comment by maruiz93** [comment]
The only CI failure is `go mod tidy` — `golang.org/x/time@v0.16.0` is being pulled in transitively and requires Go 1.26+, but CI runs Go 1.25.5 with `GOTOOLCHAIN=local`. This isn't from your changes. A rebase on latest main + `go mod tidy` should sort it out.

**Comment by lcarva** [comment]
> The only CI failure is `go mod tidy` — `golang.org/x/time@v0.16.0` is being pulled in transitively and requires Go 1.26+, but CI runs Go 1.25.5 with `GOTOOLCHAIN=local`. This isn't from your changes. A rebase on latest main + `go mod tidy` should sort it out.

@maruiz93, `go mod tidy` is currently failing on main as well. Maybe https://github.com/kubearchive/kubearchive/pull/2095 will fix it?
